### PR TITLE
feat: add dispatch core mfa plugin for signal engagement

### PR DESCRIPTION
**Pull request by ghost**
Adds a plugin for signal engagement confirmation where Dispatch authentication is the source of additional verification for an engaged user.

### CODE REVIEW

File: src/dispatch/plugins/dispatch_core/plugin.py [hunk 1]
```diff
@@ -273,6 +279,102 @@ def get(
...
+    author_url = "https://github.com/netflix/dispatch.git"
+
+    def wait_for_challenge(self, challenge_id: UUID, db_session: Session, timeout: int = 300):
+        """Waits for a multi-factor authentication challenge."""
+        import time
+
+        start_time = time.time()
+
+        while time.time() - start_time < timeout:
+            with get_organization_session(organization_slug="default") as session:
```
Comment (kevgliss) [inline src/dispatch/plugins/dispatch_core/plugin.py:301]: If this is within a web request shouldn't it already be scoped to the organization?

File: src/dispatch/plugins/dispatch_core/plugin.py [hunk 2]
```diff
@@ -273,6 +279,102 @@ def get(
...
+
+                if challenge.status == MfaChallengeStatus.APPROVED:
+                    return MfaChallengeStatus.APPROVED
+                elif challenge.status == MfaChallengeStatus.DENIED:
+                    raise Exception("Challenge denied.")
+
+            time.sleep(1)
+
+        # Update the challenge status to EXPIRED if it times out
+        with get_organization_session(organization_slug="default") as session:
```
Comment (kevgliss) [inline src/dispatch/plugins/dispatch_core/plugin.py:315]: Same question as above.

File: src/dispatch/plugins/dispatch_core/plugin.py [hunk 3]
```diff
@@ -273,6 +279,102 @@ def get(
...
+        org_slug = project.organization.slug if project.organization else "default"
+
+        challenge_url = f"{DISPATCH_UI_URL}/{org_slug}/mfa?project_id={project_id}&challenge_id={challenge.challenge_id}&action={action}"
+        return challenge, challenge_url
+
+    def validate_mfa_token(self, payload: MfaPayload, current_user: DispatchUser, db_session: Session):
+        """Validates a multi-factor authentication token."""
+        print(f"got here: {payload.challenge_id=} {payload.__dict__=}")
+        challenge = db_session.query(MfaChallenge).filter_by(challenge_id=payload.challenge_id).one_or_none()
+        print(f"disp {challenge.dispatch_user_id=} and {payload.user_id=} {type(challenge.dispatch_user_id)=} {type(payload.user_id)=}")
```
Comment (kevgliss) [inline src/dispatch/plugins/dispatch_core/plugin.py:352]: Remove or move to logging.

File: src/dispatch/static/dispatch/src/auth/Mfa.vue [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    if (!challengeId || !projectId || !action) {
+      throw new Error("Missing required parameters")
+    }
+
+    const response = await authApi.verifyMfa({
+      challenge_id: challengeId,
+      project_id: projectId,
+      action: action,
+      user_id: 12480, // TODO(whsel): fetch this
```
Comment (kevgliss) [inline src/dispatch/static/dispatch/src/auth/Mfa.vue:38]: I think this should be done server side.

**Review by kevgliss — APPROVED** [review]

**Comment by wssheldon** [comment]
Bumping..

**Pull request closed without merging**